plugin/pkg/kmsproviders/vault/vault.go

```diff
@@ -0,0 +1,240 @@
+func KMSFactory(configFile io.Reader) (envelope.Service, error) {
+	configFileContents, err := ioutil.ReadAll(configFile)
+	if err != nil {
+		return nil, fmt.Errorf("could not read contents: %v", err)
+	}
+
+	var config EnvelopeConfig
+	err = yaml.Unmarshal(configFileContents, &config)
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing file: %v", err)
+	}
+
+	err = validateConfig(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := newClientWrapper(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &vaultEnvelopeService{config: &config, client: client}, nil
+}
+
+func validateConfig(config *EnvelopeConfig) error {
+	if len(config.KeyNames) == 0 {
+		return errors.New("vault provider has no valid key names")
+	}
+
+	if config.Address == "" {
+		return errors.New("vault provider has no valid address")
+	}
+
+	return validateAuthConfig(config)
+}
+
+func validateAuthConfig(config *EnvelopeConfig) error {
+	count := 0
+
+	if config.Token != "" {
+		count++
+	}
+
+	if config.ClientCert != "" || config.ClientKey != "" {
+		if config.ClientCert == "" || config.ClientKey == "" {
+			return errors.New("vault provider has invalid TLS authentication information")
+		}
+		count++
+	}
+
+	if config.RoleID != "" || config.SecretID != "" {
+		if config.RoleID == "" {
+			return errors.New("vault provider has invalid approle authentication information")
+		}
+		count++
+	}
+
+	if count == 0 {
+		return errors.New("vault provider has no authentication information")
+	}
+	if count > 1 {
+		return errors.New("vault provider has more than one authentication information")
+	}
+
+	return nil
+}
+
+type vaultEnvelopeService struct {
+	config *EnvelopeConfig
+	client *clientWrapper
+	// We may update token for api.Client, but there is no sync for api.Client.
+	// Read lock for encrypt/decrypt requests, write lock for login requests which
+	// will update token for api.Client.
+	rwmutex sync.RWMutex
+}
+
+func (s *vaultEnvelopeService) Decrypt(data string) ([]byte, error) {
+	// Find the mached key
+	var key string
+	for _, name := range s.config.KeyNames {
+		if strings.HasPrefix(data, name+":") {
+			key = name
+			break
+		}
+	}
+	if key == "" {
+		return nil, errors.New("no matching vault key found")
+	}
+
+	// Replace the key name with "vault:" for Vault transit API
+	if !strings.HasPrefix(data, key) {
+		return nil, errors.New("encrypted data from storage does not have key prefix")
+	}
+	cipher := vaultPrefix + data[len(key):]
+
+	plain, err := s.withRefreshToken(false, key, cipher)
+	if err != nil {
+		return nil, err
+	}
+
+	return base64.StdEncoding.DecodeString(plain)
+}
+
+func (s *vaultEnvelopeService) Encrypt(data []byte) (string, error) {
+	// Use the frist key to encrypt
+	key := s.config.KeyNames[0]
+	plain := base64.StdEncoding.EncodeToString(data)
+
+	cipher, err := s.withRefreshToken(true, key, plain)
+	if err != nil {
+		return "", err
+	}
+
+	// The format of cipher from Vault is "vault:v1:....".
```

Comment:
Why are you making "vault:foo" into "key:foo" instead of "key:vault:foo" or "vault:key:foo"?  It makes this feel lossy.

Reply:
"vault" prefix is added by the enveloper transformer. so the overall encoding does include "vault" to identify the provider if more than one is configured.
overall encoding:
k8s:enc:kms:<api-version>:vault:len(<KEK-key-name>:<KEK-key-version>:<DEK
encrypted with KEK>):<KEK-key-name>:<KEK-key-version>:<DEK encrypted with KEK>